kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/stable/StableIdents.kt

```diff
@@ -0,0 +1,133 @@
+    * If the object is annotated with [IsStableType] then the toString method is used.
+    * If the object is an instance of [WithDataTestName] then the [WithDataTestName.dataTestName] method is used.
+    * If the object is considered stable via [isStable] then the toString method is used.
+    *
+    * Otherwise, the class name is returned.
+    */
+   fun getStableIdentifier(t: Any?): String {
+      return when {
+         t == null -> "<null>"
+         t::class.hasAnnotation<IsStableType>() || platform != Platform.JVM -> t.toString()
+         t is WithDataTestName -> t.dataTestName()
+         else -> {
+            val psv = platformStableValue(t)
+            return when {
+               psv != null -> psv
+               isStable(t::class, t) -> t.toString()
+               else -> t::class.bestName()
+            }
+         }
+      }
+   }
+
+   /**
+    * Returns true if the given type is a class and stable.
+    */
+   private fun isStable(type: KType) = when (val classifier = type.classifier) {
+      is KClass<*> -> isStable(classifier, null)
+      else -> false
+   }
+
+   /**
+    * Returns true if the given class is a "stable" class, ie one that has a consistent toString().
+    *
+    *      An instance is considered stable if it is a data class where each parameter is either a data class itself,
+    *      one of the [allPlatformStableTypes], or if the type of instance is annotated with [IsStableType].
+    *
+    *      Note: If the user has overridden `toString()` and the returned value is not stable,
+    *      then the behavior of this function is undefined.
+    *
+    * A stable class is either:
+    *  - a Kotlin or Java primitive type or a String
+    *  - an enum class
+    *  - a data class that only contains stable classes as members
+    *  - a stable type on the executing platform (see [isPlatformStable]).
+    *  - a collection type where the elements are stable
+    */
+   internal fun isStable(kclass: KClass<*>, t: Any? = null): Boolean {
+      return when {
+         isAllPlatformStable(kclass) -> true
+         reflection.isEnumClass(kclass) -> true
+         reflection.isDataClass(kclass) && hasStableMembers(kclass, t) -> true
+         isPlatformStable(kclass) -> true
+         else -> {
+            println("Warning, type $kclass used in data testing does not have a stable toString()")
+            false
+         }
+      }
+   }
+
+   internal fun isStableCollection(kclass: KClass<*>): Boolean {
+      return when (kclass) {
+         Map::class -> true
+         else -> false
+      }
+   }
+
+   /**
+    * Returns true if all members of the given [kclass] are "stable".
+    */
+   private fun hasStableMembers(kclass: KClass<*>, t: Any? = null) =
+      reflection.primaryConstructorMembers(kclass).let { members ->
+         members.isNotEmpty() && members.all { getter ->
+            val typeIsStable = isStable(getter.type)
+            if (t == null) {
+               typeIsStable
```

Comment:
my knee-jerk reaction is this: "if something is `null`, it surely is stable". What am I missing?